crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,749 @@
+///
+/// Note that this does not compare mathematical equality, but data equality. This means
+/// that two observables might represent the same observable but not compare as equal.
+///
+/// @param obs A pointer to one observable.
+/// @param other A pointer to another observable.
+///
+/// @return ``true`` if the observables are equal, ``false`` otherwise.
+///
+/// # Example
+///
+///     QkSparseObservable *observable = qk_obs_identity(100);
+///     QkSparseObservable *other = qk_obs_identity(100);
+///     bool are_equal = qk_obs_equal(observable, other);
+///
+/// # Safety
+///
+/// Behavior is undefined if ``obs`` or ``other`` are not valid, non-null pointers to
+/// ``QkSparseObservable``\ s.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_equal(
+    obs: *const SparseObservable,
+    other: *const SparseObservable,
+) -> bool {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    let other = unsafe { const_ptr_as_ref(other) };
+
+    obs.eq(other)
+}
+
+/// @ingroup QkSparseObservable
+/// Print the observable.
+///
+/// @param obs A pointer to the ``SparseObservable`` to print.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     qk_obs_print(obs);
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_print(obs: *const SparseObservable) {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    println!("{:?}", obs);
+}
+
+/// @ingroup QkSparseTerm
+/// Print the sparse term.
+///
+/// @param term A pointer to the term.
+///
+/// @return The function exit code. This is ``>0`` if reading the term failed.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     QkSparseTerm term;
+///     qk_obs_term(obs, &term, 0);
+///     int exit_code = qk_obsterm_print(&term);
+///
+/// # Safety
+///
+/// Behavior is undefined ``term`` is not a valid, non-null pointer to a ``QkSparseTerm``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obsterm_print(term: *const CSparseTerm) -> ExitCode {
+    let term = unsafe { const_ptr_as_ref(term) };
+    let view: SparseTermView = match term.try_into() {
+        Ok(view) => view,
+        Err(err) => return ExitCode::from(err),
+    };
+
+    println!("{:?}", view);
+    ExitCode::Success
+}
+
+/// @ingroup QkBitTerm
+/// Get the label for a bit term.
+///
+/// @param bit The bit term.
+///
+/// @return The label as unsigned integer.
+///
+/// # Example
+///     
+///     QkBitTerm bit_term = QkBitTerm_Y;
+///     // cast the uint32_t to char
+///     char label = qk_bitterm_label(bit_term);
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_bitterm_label(bit_term: BitTerm) -> u32 {
+    // BitTerm is implemented as u8, which is calling convention compatible with C,
+    // hence we can pass ``bit_term`` by value
```

Comment:
We can pass `BitTerm` by value, but this is super unsafe if C passes us a `u8` that isn't a valid representation of a `BitTerm` lol.  That's part of the larger conversation about how much error-checking we want to do, though.

Reply:
We can add some error checking -- or can we argue that the signature expects a `BitTerm` (which is a defined enum) so passing something else is undefined? Or doesn't that work because of the `typedef uint_8 QkBitTerm`?

Reply:
I think the safety comment you added is fine for now - it's clearly documenting that it's undefined behaviour to pass an invalid value.